Sources/Kitura-CredentialsTwitter/CredentialsTwitter.swift

```diff
@@ -0,0 +1,358 @@
+    ///   - response: The `RouterResponse` object used to respond to the request.
+    ///   - onFailure: The closure to invoke in the case of an authentication failure.
+    ///   - onPass: The closure to invoke when the plugin doesn't recognize the authentication data (usually an
+    ///       authentication token) in the request.
+    ///   - inProgress: The closure to invoke to cause a redirect to the login page in the case of redirecting
+    ///       authentication.
+    func twitterTokenRequest(request: RouterRequest,
+                             response: RouterResponse,
+                             onFailure: @escaping (HTTPStatusCode?, [String:String]?) -> Void,
+                             onPass: @escaping (HTTPStatusCode?, [String:String]?) -> Void,
+                             inProgress: @escaping () -> Void) {
+        var requestOptions: [ClientRequest.Options] = []
+        requestOptions.append(.schema("https://"))
+        requestOptions.append(.hostname("api.twitter.com"))
+        requestOptions.append(.method("POST"))
+        requestOptions.append(.path("oauth/request_token"))
+        var headers = [String : String]()
+        var parameters = [String:String]()
+
+        let nonce = String.nonce
+        let timeStamp = Date().timeIntervalSince1970.roundedString
+
+        parameters["oauth_consumer_key"] = consumerKey
+        parameters["oauth_signature_method"] = "HMAC-SHA1"
+        parameters["oauth_timestamp"] = timeStamp
+        parameters["oauth_nonce"] = nonce
+        parameters["oauth_version"] = "1.0"
+
+        guard let signature = String.oAuthSignature(fromMethod: "POST",
+                                                    urlString: "https://api.twitter.com/oauth/request_token",
+                                                    parameters: parameters,
+                                                    consumerSecret: consumerSecret) else {
+                                                        onFailure(nil, nil)
+                                                        return
+        }
+
+        parameters["oauth_signature"] = signature
+
+        headers["Authorization"] = String.oAuthAuthorizationString(fromParameters: parameters)
+        requestOptions.append(.headers(headers))
+
+        let twitterRequest = HTTP.request(requestOptions) {
+            (optionalResponse) in
+            guard let tokenResponse = optionalResponse, tokenResponse.statusCode == HTTPStatusCode.OK else {
+                Log.error("Twitter token response returned with \(optionalResponse?.statusCode)")
+                onFailure(optionalResponse?.statusCode, nil)
+                return
+            }
+
+            var body = Data()
+            do {
+                try tokenResponse.readAllData(into: &body)
+                let string = String(data: body, encoding: String.Encoding.utf8)
+                guard let fields = string?.components(separatedBy: "&") else {
```

Comment:
URLComponents isn't an option, consider pulling out the parsing into its own method so it's not repeated